joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -84,31 +85,55 @@ JointTrajectoryController::update()
+  if (traj_point_active_ptr_ && (*traj_point_active_ptr_)->is_empty() == false) {
+    // find next new point for current timestamp
+    auto traj_point_ptr = (*traj_point_active_ptr_)->sample(lifecycle_node_->now());
+    // find next new point for current timestamp
+    // set cmd only if a point is found
+    if (traj_point_ptr == (*traj_point_active_ptr_)->end()) {
+      return CONTROLLER_INTERFACE_RET_SUCCESS;
+    }
 
-  // find next new point for current timestamp
-  auto traj_point_ptr = (*traj_point_active_ptr_)->sample(rclcpp::Clock().now());
-  // find next new point for current timestamp
-  // set cmd only if a point is found
-  if (traj_point_ptr == (*traj_point_active_ptr_)->end()) {
-    return CONTROLLER_INTERFACE_RET_SUCCESS;
-  }
+    // check if new point ptr points to the same as previous point
+    if (prev_traj_point_ptr_ == traj_point_ptr) {
+      return CONTROLLER_INTERFACE_RET_SUCCESS;
+    }
 
-  // check if new point ptr points to the same as previous point
-  if (prev_traj_point_ptr_ == traj_point_ptr) {
-    return CONTROLLER_INTERFACE_RET_SUCCESS;
-  }
+    size_t joint_num = registered_joint_cmd_handles_.size();
+    for (size_t index = 0; index < joint_num; ++index) {
+      registered_joint_cmd_handles_[index]->set_cmd(traj_point_ptr->positions[index]);
+    }
 
-  size_t joint_num = registered_joint_cmd_handles_.size();
-  for (size_t index = 0; index < joint_num; ++index) {
-    registered_joint_cmd_handles_[index]->set_cmd(traj_point_ptr->positions[index]);
+    prev_traj_point_ptr_ = traj_point_ptr;
+    set_op_mode(hardware_interface::OperationMode::ACTIVE);
   }
 
-  prev_traj_point_ptr_ = traj_point_ptr;
-  set_op_mode(hardware_interface::OperationMode::ACTIVE);
+  // publish state
+  {
+    if (state_publisher_period_.seconds() > 0.0 &&
+      (last_state_publish_time_ + state_publisher_period_) < lifecycle_node_->now())
+    {
+      if (state_publisher_ && state_publisher_->trylock()) {
+        last_state_publish_time_ = lifecycle_node_->now();
+
+        state_publisher_->msg_.header.stamp = last_state_publish_time_;
+        // TODO(ddengster): The following code is ported from ros_controllers.
+        // Port it when we're ready to put in the other trajectory code.
+
+        // state_publisher_->msg_.header.stamp          = time_data_.readFromRT()->time;
+        // state_publisher_->msg_.desired.positions     = desired_state_.position;
+        // state_publisher_->msg_.desired.velocities    = desired_state_.velocity;
+        // state_publisher_->msg_.desired.accelerations = desired_state_.acceleration;
+        // state_publisher_->msg_.actual.positions      = current_state_.position;
+        // state_publisher_->msg_.actual.velocities     = current_state_.velocity;
+        // state_publisher_->msg_.error.positions       = state_error_.position;
+        // state_publisher_->msg_.error.velocities      = state_error_.velocity;
+
+        state_publisher_->unlockAndPublish();
```

Comment:
Do we still need this to be scoped if we're manually holding/releasing the mutex?

Reply:
Yes, we don't need it. This scoping is a code style thing. See [reply](https://github.com/ros-controls/ros2_controllers/pull/25#discussion_r386767757) above.